### DISCUSSION TITLE
How do I get the "raw" source code of an AST node in a plugin?

### DISCUSSION THREAD

**Discussion by ole-treichel** [swc-project/swc#4961] [Q&A] [answered]
I am trying to implement something similiar to this [babel plugin](https://github.com/dogma-io/babel-plugin-react-code-block) as swc-plugin. You can configure a component that will receive the raw, uncompiled JSX markup of its children as a prop, e.g.:

Input:
```javascript
<Demo>
  <Button>Hello</Button>
</Demo>
```

Output:
```javascript
<Demo code={`<Button>Hello</Button>`}>
  <Button>Hello</Button>
</Demo>
```

I noticed that every AST-node has a `span` field, but I am unsure how I would use this to get to the corresponding source code. 

I am wondering if this is possible at all, maybe I am misunderstanding what the `span` field is for

**Comment by kwonoj** [comment, marked as answer]
You may able to try few interfaces around `SourceMapper`, like https://github.com/swc-project/swc/blob/main/crates/swc_plugin_proxy/src/source_map/plugin_source_map_proxy.rs#L175=

**Reply by ole-treichel** [reply]
Yepp that works, thanks 👍 .

One follow up question (Hope this is okay): 

I see that the `PluginSourceMapProxy` is only implemented for the `wasm32` target. If I understand correctly, `swc_ecma_transforms_testing` tests plugins without compiling them to `wasm`. Is there a way to use `swc_ecma_transforms_testing` in a way that uses the plugin compiled to `wasm` or supports `PluginSourceMapProxy` in any other way?

**Reply by kwonoj** [reply]
you need / should use `SourceMap` for the non plugin context instead of plugin proxy. That's the reason plugin passes trait `SourceMapper` instead of implementation to allow to pass correct implementation depends on the context.
